### PR TITLE
Added container error handling in PreRun/PostRun

### PR DESCRIPTION
**Pull request by bdevcich-hpe**
Signed-off-by: Blake Devcich <blake.devcich@hpe.com>

### CODE REVIEW

File: docs/guides/user-containers/readme.md [hunk 1]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
     GFS2 filesystems have special considerations since the mount directory contains directories for
     every compute node. See [GFS2 Index Mounts](#gfs2-index-mounts) for more info.
 
+### Targeting Nodes
+
+For container directives, compute nodes must be assigned to the workflow. The NNF software will
+trace the compute nodes back to their local NNF nodes and the containers will be executed on those
+NNF nodes. The act of assigning compute nodes to your container workflow instructs the NNF software
+to select the NNF nodes that run the containers.
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:270]: This last sentence seems redundant with the prior sentence.

File: docs/guides/user-containers/readme.md [hunk 2]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+
+For the `jobdw` directive that is included above, the servers (i.e. NNF nodes) must also be assigned
+along with the computes.
+
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:274]: If a compute node not associated with any of the Rabbits assigned in the Servers object are specified, what happens?

File: docs/guides/user-containers/readme.md [hunk 3]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
 ## Running a Container Workflow
 
-TODO: Once the container exit strategy is complete, come back to this section to detail:
+Once the workflow is created, it will need to progress through the following states. This is a quick
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:277]:
```suggestion
Once the workflow is created, the WLM progresses it through the following states. This is a quick
```

File: docs/guides/user-containers/readme.md [hunk 4]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+overview of the container-related behavior that occurs:
+
+- Proposal: Verify storages are provided according to the container profile.
+- Setup: If applicable, request ports from NnfPortManager.
+- DataIn: No container related activity.
+- PreRun: Appropriate `MPIJob` or `Job(s)` are created for the workflow. In turn, user containers
+are created and launched by Kubernetes. Containers are expected to have started in this state.
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:284]:
```suggestion
are created and launched by Kubernetes. Containers are expected to start in this state.
```

File: docs/guides/user-containers/readme.md [hunk 5]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+- PostRun: Once in PostRun, user containers are expected to have completed (non-zero exit)
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:285]:
```suggestion
- PostRun: Once in PostRun, user containers are expected to complete (non-zero exit)
```

File: docs/guides/user-containers/readme.md [hunk 6]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+- DataOut: No container related activity.
+- Teardown: Ports are released; `MPIJob` or `Job(s)` are deleted, which in turn deletes the user
+containers.
+
+The two main states of a container workflow (i.e. PreRun, PostRun) are discussed further in the
+following sections.
+
+### PreRun
+
+In PreRun, the containers are created and expected to start. Once the containers have reach a
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:296]:
```suggestion
In PreRun, the containers are created and expected to start. Once the containers reach a
```

File: docs/guides/user-containers/readme.md [hunk 7]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+non-initialization state (i.e. Running), the containers are considered to have been started and the
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:297]:
```suggestion
non-initialization state (i.e. Running), the containers are considered to be started and the
```

File: docs/guides/user-containers/readme.md [hunk 8]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+Containers into the container specification for various reasons:
 
-This will mostly be a from a WLM/DWS Workflow integration perspective, but will touch on the lower
-level workings.
+- To ensure the proper permissions (i.e. user/group) are configured
+- For MPI jobs, to ensure the launcher pod can contact the worker pods
+
+### PostRun
+
+In PostRun, the containers are expected to exit cleanly with a zero exit code. If a container does
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:323]:
```suggestion
In PostRun, the containers are expected to exit cleanly with a zero exit code. If a container fails to
```

File: docs/guides/user-containers/readme.md [hunk 9]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+not exit cleanly, the Kubernetes software will attempt a number of retries based on the
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:324]:
```suggestion
exit cleanly, the Kubernetes software attempts a number of retries based on the
```

File: docs/guides/user-containers/readme.md [hunk 10]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+configuration of the container profile. It will continue to do this until the container exits
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:325]:
```suggestion
configuration of the container profile. It continues to do this until the container exits
```

File: docs/guides/user-containers/readme.md [hunk 11]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+successfully, or if the `retryLimit` is set - whichever occurs first. In the latter case, the
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:326]:
```suggestion
successfully, or until the `retryLimit` is hit - whichever occurs first. In the latter case, the
```

File: docs/guides/user-containers/readme.md [hunk 12]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+workflow will report an Error.
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:327]:
```suggestion
workflow reports an Error.
```

File: docs/guides/user-containers/readme.md [hunk 13]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+- If the container exits unsuccessfully after `retryLimit` number of retries, transition to the
+Error State.
+- If the container is running and has not exited after `postRunTimeoutSeconds` seconds, terminate
+the container and transition to the Error State.
+
+### Failure Retries
+
+If a container fails (non-zero exit code), the Kubernetes software implements retries. The number of
+retries can be set via the `retryLimit` field in the container profile. If a non-zero exit code is
+detected, the Kubernetes software will create a new instance of the pod and try again. The
```
Comment (ajfloeder) [inline docs/guides/user-containers/readme.md:348]:
```suggestion
detected, the Kubernetes software creates a new instance of the pod and retries. The
```

File: docs/guides/user-containers/readme.md (hunk 2 above)
Comment (bdevcich-hpe) [reply docs/guides/user-containers/readme.md:274]: Good question. The NNF file system wouldn't be available on the rabbit and therefore the mount directory passed into the container won't exist. I'll look into this more and add any fixes, if necessary.

Comment (ajfloeder) [reply docs/guides/user-containers/readme.md:274]: Maybe the webhook should check this?

**Review by ajfloeder — APPROVED** [review]

File: docs/guides/user-containers/readme.md [hunk 14]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+For the `jobdw` directive that is included above, the servers (i.e. NNF nodes) must also be assigned
+along with the computes.
+
 ## Running a Container Workflow
 
-TODO: Once the container exit strategy is complete, come back to this section to detail:
+Once the workflow is created, the WLM progresses it through the following states. This is a quick
+overview of the container-related behavior that occurs:
+
+- Proposal: Verify storages are provided according to the container profile.
```
Comment (roehrich-hpe) [inline docs/guides/user-containers/readme.md:280]: Can "storages" contain a hyperlink up to the "Container Storages" section?  I still stumble on some of the short-hand lingo.

File: docs/guides/user-containers/readme.md [hunk 15]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+terminate the container and transition to the Error State.
+
+#### Init Containers
 
-- PreRun Behavior
-- PostRun Behavior
-- Teardown
+Containers have some additional initialization that occurs using [Init
+Containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/).  These containers
+must run to completion before the main container can start. The NNF Software injects Init
```
Comment (roehrich-hpe) [inline docs/guides/user-containers/readme.md:315]: This:
"The NNF Software injects Init Containers into the container specification"
is the key statement in this paragraph, and sub-section.  It should be the first sentence.

File: docs/guides/user-containers/readme.md [hunk 16]
```diff
@@ -262,16 +262,106 @@ allowed per workflow.
...
+It is important to note that due to the configuration of the `MPIJob` and/or `Job` that is created
+for User Containers, the container retries are immediate - there is no backoff timeout between
+retires. This is due to the NNF Software setting the `RestartPolicy` to `Never`, which causes a new
+pod to spin up after every failure rather than re-use (i.e. restart) the previously failed pod. This
+allows a user to see a complete history of the failed pod(s) and the logs can easily be obtained.
+See more on this at [Handling Pod and container
+failures](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
+in the Kubernetes documentation.
 
 ## Putting it All Together
```
Comment (roehrich-hpe) [inline docs/guides/user-containers/readme.md:366]: I'm missing a piece: How does the user container know when PostRun begins?  Where does it get that cue?

Comment (bdevcich-hpe) [reply docs/guides/user-containers/readme.md:366]: Is this regarding the PostRun timeout? The container has no idea - the NNF software sets the Job's activeDeadline when the workflow transitions to PostRun. Once that activeDeadline timeout is hit, k8s takes care of the rest and the containers are killed. The job transitions to a failed state and the NNF software turns that into a workflow error.

File: docs/guides/user-containers/readme.md (hunk 15 above)
Comment (bdevcich-hpe) [reply docs/guides/user-containers/readme.md:315]: I'll reword this

File: docs/guides/user-containers/readme.md (hunk 14 above)
Comment (bdevcich-hpe) [reply docs/guides/user-containers/readme.md:280]: Good call

**Review by ajfloeder — APPROVED** [review]

File: docs/guides/user-containers/readme.md (hunk 16 above)
Comment (roehrich-hpe) [reply docs/guides/user-containers/readme.md:368]: Is the user container supposed to do some work during PostRun?

Comment (bdevcich-hpe) [reply docs/guides/user-containers/readme.md:368]: That's up to the application developer that created the image, but they won't have any cue from the NNF software that it's in PostRun.

I envision the port being the mechanism that drives behavior to the application (e.g. SCR) running in the container. The compute node application contacts the application in the container and tells it that it's time to do something.

1. In PreRun, user container starts successfully. The workflow transitions to PreRun complete
2. Flux runs the application on the compute node
3. The compute node application contacts the user container via the port to drive behavior, ultimately saying: "I'm done"
4. Compute node application finishes and Flux sees that
5. Flux transitions to PostRun
6. PostRun expects the user container to exit cleanly before transitioning to PostRun complete

--- 

@roehrich-hpe and I discussed this at my desk already, but posting this for posterity.

I'll also be adding a section that illustrates this more directly.

File: docs/guides/user-containers/readme.md [hunk 17]
```diff
@@ -262,16 +262,119 @@ allowed per workflow.
...
-- Teardown
+- Ensuring the proper permissions (i.e. UID/GID) are available in the main container
+- For MPI jobs, ensuring the launcher pod can contact each worker pod via DNS
 
-This will mostly be a from a WLM/DWS Workflow integration perspective, but will touch on the lower
-level workings.
+### PreRun Completed
+
+Once PreRun has transitioned to `Completed` status, the user container is now running.
+Sequentially, the WLM should initiate applications on the compute nodes. Utilizing [container
```
Comment (matthew-richerson) [inline docs/guides/user-containers/readme.md:326]: What do you mean by "sequentially" here? The user apps on the compute nodes should be able to start at the same time.

**Review by matthew-richerson — APPROVED** [review]

File: docs/guides/user-containers/readme.md (hunk 17 above)
Comment (bdevcich-hpe) [reply docs/guides/user-containers/readme.md:326]: I was making the assumption that Flux would wait for Completed before starting the application on the compute nodes. Maybe I should reword?

**Review by roehrich-hpe — APPROVED** [review]

File: docs/guides/user-containers/readme.md (hunk 17 above)
Comment (matthew-richerson) [reply docs/guides/user-containers/readme.md:326]: I read it as "start the application on compute nodes one at a time." Maybe that's just me.

Comment (bdevcich-hpe) [reply docs/guides/user-containers/readme.md:326]: I can see that. I'll reword this.

**Pull request merged**